src/App.ts

```diff
@@ -516,10 +521,31 @@ export default class App<AppCustomContext extends StringIndexed = StringIndexed>
   /**
    * Register CustomFunction middleware
    */
-  public function(callbackId: string, ...listeners: CustomFunctionMiddleware): this {
-    const fn = new CustomFunction(callbackId, listeners, this.webClientOptions);
-    const m = fn.getMiddleware();
-    this.middleware.push(m);
+  public function<Options extends SlackEventMiddlewareArgsOptions = { autoAcknowledge: true }>(
+    callbackId: string,
+    options: Options,
+    ...listeners: Middleware<SlackCustomFunctionMiddlewareArgs<Options>>[]
+  ): this;
+  public function<Options extends SlackEventMiddlewareArgsOptions = { autoAcknowledge: true }>(
+    callbackId: string,
+    ...listeners: Middleware<SlackCustomFunctionMiddlewareArgs<Options>>[]
+  ): this;
+  public function<Options extends SlackEventMiddlewareArgsOptions = { autoAcknowledge: true }>(
+    callbackId: string,
+    ...optionOrListeners: (Options | Middleware<SlackCustomFunctionMiddlewareArgs<Options>>)[]
+  ): this {
+    // TODO: fix this casting; edge case is if dev specifically sets AutoAck generic as false, this true assignment is invalid according to TS.
+    const options = isSlackCustomFunctionMiddlewareArgsOptions<Options>(optionOrListeners[0])
+      ? optionOrListeners[0]
+      : ({ autoAcknowledge: true } as Options);
+    const listeners = optionOrListeners.filter(
+      (optionOrListener): optionOrListener is Middleware<SlackCustomFunctionMiddlewareArgs<Options>> => {
+        return !isSlackCustomFunctionMiddlewareArgsOptions<Options>(optionOrListener);
+      },
+    );
+
+    const fn = new CustomFunction<Options>(callbackId, listeners, options);
+    this.listeners.push(fn.getListeners());
```

Comment:
What's the impact of pushing this into the `listeners` array vs. the previous `middleware` array?

Reply:
From my testing I have not found any impact 🤔 

Would like to know what @seratch thinks about this

Reply:
@WilliamBergamin @filmaj If an app has a few custom middleware before/after this function middleware registration, the timing of the execution of this function stuff can be changed. I'd say it is a minor behavior difference. See https://github.com/slackapi/bolt-js/blob/%40slack/bolt%403.22.0/src/App.ts#L1133-L1180 to learn how the this.middleware and this.listeners are executed when receiving an event request.

Thinking of the appropriate behavior, changing this way sounds more intuitive to me because the function method receives "listeners", not a global middleware. My sugguestion is to add TODO comment this time, and then apply the behavior change in v4.

Reply:
Or if we're aiming to add this in v4, mentioning the behavior change in the migration guide would be enough

Reply:
OK, sounds good. One thought here: we should set aside some time to experiment with adding a full-fledged bolt custom function interactivity end to end test to platform-devxp-test, and look to integrate those checks into CI. It would give us a lot more confidence that no edge cases break (and lets us rely less on manual testing).